WreckEngine/gl_structs.h

```diff
@@ -14,231 +14,381 @@
-	shared<GLuint> program{ new GLuint(def), local(delShaderProg) };
-	inline GLuint& operator()() const { return *program; }
-
-	inline void create() { if (*program != def) return; *program = glCreateProgram(); }
-	// properly sets up the program once the shaders are set
-	inline void link() const { 
-		if (vertex.valid())   glAttachShader(*program, vertex());
-		if (fragment.valid()) glAttachShader(*program, fragment());
-		glLinkProgram(*program);
-	}
-
-	inline void use() const { glUseProgram(*program); }
-	// used for retrieving uniform variables (of type T)
-	template<typename T> inline GLuniform<T> getUniform(const char* name) const { GLuniform<T> u; u.location = glGetUniformLocation(*program, name); return u; }
+    GLshader vertex, fragment;// this can be extended when necessary for geometry and tessellation shaders, or use a std::vector
+    shared<GLuint> program {new GLuint(def), local(delShaderProg)};
+    inline GLuint& operator()() const {
+        return *program;
+    }
+    inline operator bool() const {
+        return *program != def;
+    }
+
+    inline void create() {
+        if (*program != def) return;
+        GL_CHECK(*program = glCreateProgram());
+    }
+    // properly sets up the program once the shaders are set
+    inline void link() const {
+        if (vertex.valid())   GL_CHECK(glAttachShader(*program, vertex()));
+        if (fragment.valid()) GL_CHECK(glAttachShader(*program, fragment()));
+        GL_CHECK(glLinkProgram(*program));
+    }
+
+    inline void use() const {
+        GL_CHECK(glUseProgram(*program));
+    }
+    // used for retrieving uniform variables (of type T)
+    template<typename T> inline GLuniform<T> getUniform(const char* name) const {
+        GLuniform<T> u;
+        GL_CHECK(u.location = glGetUniformLocation(*program, name));
+        return u;
+    }
+};
+
+/**
+ * Defines an OpenGL rendering state.
+ */
+class GLstate {
+public:
+    /**
+     * \brief Creates a new OpenGL state.
+     */
+    GLstate();
+
+    /**
+     * \brief Gets the OpenGL state at the top of the stack.
+     *
+     * \return The OpenGL state at the top of the stack, or null if there are no states.
+     */
+    static GLstate* peek();
+
+    /**
+     * \brief Pushes the current OpenGL state.
+     */
+    static void push();
+
+    /**
+     * \brief Attempts to pop an OpenGL state off of the stack.
+     *
+     * \return True if there was a state to pop, otherwise false.
+     */
+    static bool pop();
+
+private:
+    static std::vector<GLstate> s_States;
+    GLint m_Viewport[4];
+    GLint m_ScissorBox[4];
+    GLint m_Program;
+    GLint m_Texture;
+    GLint m_ActiveTexture;
+    GLint m_ArrayBuffer;
+    GLint m_ElementArrayBuffer;
+    GLint m_VertexArray;
+    GLint m_BlendSrc;
+    GLint m_BlendDst;
+    GLint m_BlendEquationRGB;
+    GLint m_BlendEquationAlpha;
+    GLboolean m_EnableBlend;
+    GLboolean m_EnableCullFace;
+    GLboolean m_EnableDepthTest;
+    GLboolean m_EnableScissorTest;
```

Comment:
the naming practice where you prefix everything by their letter is a thing, I'd prefer we only resort to it when necessary. My convention is typically to drop to this after underscores alone become too inconvenient. In this case, it's definitely more appropriate to use the base names imo. Also, given the sheer number of variables here I'd use multiple declarations on the same line to make it more readable. (a little paradoxical I know)